grudge/reductions.py

```diff
@@ -80,19 +80,21 @@ def _norm(dcoll: DiscretizationCollection, vec, p, dd):
         return np.fabs(vec)
     if p == 2:
         from grudge.op import _apply_mass_operator
-        return abs(
-            nodal_sum(
-                dcoll,
-                dd,
-                vec.conj() * _apply_mass_operator(dcoll, dd, dd, vec))
-            )**(1/2)
+        return vec.array_context.np.sqrt(
+            # Quantities being summed are real up to rounding error, so abs() can
+            # go on the outside
```

Comment:
This isn't just for real numbers though, since there's a `conj` in there. So, it would seem better to either
* remove the `conj` and assert that `vec.entry_dtype` is real
* add back the `abs` to handle complex numbers too

What do you think?

Reply:
Note that I didn't actually change the `abs` here, I just added a comment to attempt to explain why it (apparently) works even though the `abs` looks like it's in the wrong place (outside the sum vs. inside).

Reply:
If I understand correctly, the result of `vec.conj() * _apply_mass_operator(dcoll, dd, dd, vec)` would be complex-typed but nonnegative real-valued (up to floating point error), so doing the `abs()` on the inside or on the outside should give the same result (also up to floating point error). Is this right @thomasgibson?

Reply:
@majosm You are correct, though if we do decide to change our basis functions (for example using Berstein polynomials), the mass operation is no long SPD (I think) and hence you're not guaranteed to have positiveness. In which case, the correct approach is to put `abs` _inside_. So I think if we want to have something more bullet-proof, let's put `abs` inside.

Does that sound okay to you guys?